### ISSUE TITLE
Variable of Optional type inside isinstance check will become Never inside for loop

### ISSUE THREAD

**Issue by falan** [microsoft/pyright#691]
**Describe the bug**
Operator "+" not supported for types "Never" and "Literal['hello']" inside for loop

**To Reproduce**
```python
from typing import Optional

never_var: Optional[str] = None
for i in ["hello", "world"]:
    if isinstance(i, str):
        if never_var is None:
            never_var = "hello"
        else:
            never_var = never_var + ", hello again"
```

**Expected behavior**
No Error.

**VS Code extension or command-line**
Running pyright 1.1.39 in VS Code on Windows 10.

**Comment by erictraut** [comment]
Thanks for the clear and concise bug report. This was a problem in how the code flow engine handled type dependencies within loops. The solution was quite involved, but the problem is now fixed. It will be included in the next version of Pyright.

**Comment by erictraut** [comment]
This is now fixed in Pyright 1.1.40, which I just published.
